awx_collection/plugins/module_utils/tower_api.py

```diff
@@ -61,370 +27,80 @@ def __init__(self, argument_spec, **kwargs):
-            # Sanity check: Did the server send back some kind of internal error?
-            if he.code >= 500:
-                self.fail_json(msg='The host sent back a server error ({1}): {0}. Please check the logs and try again later'.format(self.url.path, he))
-            # Sanity check: Did we fail to authenticate properly?  If so, fail out now; this is always a failure.
-            elif he.code == 401:
-                self.fail_json(msg='Invalid Tower authentication credentials for {0} (HTTP 401).'.format(self.url.path))
-            # Sanity check: Did we get a forbidden response, which means that the user isn't allowed to do this? Report that.
-            elif he.code == 403:
-                self.fail_json(msg="You don't have permission to {1} to {0} (HTTP 403).".format(self.url.path, method))
-            # Sanity check: Did we get a 404 response?
-            # Requests with primary keys will return a 404 if there is no response, and we want to consistently trap these.
-            elif he.code == 404:
-                if kwargs.get('return_none_on_404', False):
-                    return None
-                self.fail_json(msg='The requested object could not be found at {0}.'.format(self.url.path))
-            # Sanity check: Did we get a 405 response?
-            # A 405 means we used a method that isn't allowed. Usually this is a bad request, but it requires special treatment because the
-            # API sends it as a logic error in a few situations (e.g. trying to cancel a job that isn't running).
-            elif he.code == 405:
-                self.fail_json(msg="The Tower server says you can't make a request with the {0} method to this endpoing {1}".format(method, self.url.path))
-            # Sanity check: Did we get some other kind of error?  If so, write an appropriate error message.
-            elif he.code >= 400:
-                # We are going to return a 400 so the module can decide what to do with it
-                page_data = he.read()
-                try:
-                    return {'status_code': he.code, 'json': loads(page_data)}
-                # JSONDecodeError only available on Python 3.5+
-                except ValueError:
-                    return {'status_code': he.code, 'text': page_data}
-            elif he.code == 204 and method == 'DELETE':
-                # A 204 is a normal response for a delete function
-                pass
-            else:
-                self.fail_json(msg="Unexpected return code when calling {0}: {1}".format(self.url.geturl(), he))
-        except(Exception) as e:
-            self.fail_json(msg="There was an unknown error when trying to connect to {2}: {0} {1}".format(type(e).__name__, e, self.url.geturl()))
-        finally:
-            self.url = self.url._replace(query=None)
+            return self.tower_connector.get_one(endpoint, *args, **kwargs)
+        except AnsibleTowerException as ate:
+            self.exit_tower_exception(ate)
```

Comment:
This is still a lot of repetition of the error handling, which is something people specifically didn't like about the old modules. The only reason I can see for handling the exception _each time_ is to add contextual information, like "got a 401 while looking up related organization foo", which it doesn't do.

I am still split myself whether this should be done via a callback passed to `__init__` or a shared method by the calling class. Either would be fine.

The way the classes have been split here, consolidating the `try-except` code can't be done well, because you are keeping around multiple types of methods like `get_one` as opposed to separating responsibilities on `make_request`. You could do the error handling callback instead and be able to keep around all these utility methods in the connector class.. but you shouldn't. The connector should have a super duper stripped down interface, and expose as little surface area to other code as possible.

I've strongly landed on splitting the interface on that `make_request` point. Things like `get_one` are highly specific module utilities. They should never be used by the inventory plugin, lookup plugin, or any other plugin I can imagine.

Reply:
Altered to only have make_request, login and logout in the tower_connector.